src/oracle/connection/raw.rs

```diff
@@ -87,41 +105,62 @@ fn parse_db_string(database_url: &str) -> ConnectionResult<(String, String, Stri
         let msg = format!("Could not use {} with oci backend", database_url);
         return Err(ConnectionError::InvalidConnectionUrl(msg));
     }
-    let mut state = ParseState::UserName;
-    let mut user = String::new();
-    let mut password = String::new();
-    let mut db_url = String::from("//");
-    for c in database_url[6..].chars() {
-        match state {
-            ParseState::UserName => {
-                if c == ':' {
-                    state = ParseState::Password;
-                } else {
-                    user.push(c);
-                }
-            }
-            ParseState::Password => {
-                if c == '@' {
-                    state = ParseState::ConnectionString;
-                } else {
-                    password.push(c);
-                }
-            }
-            ParseState::ConnectionString => db_url.push(c),
-        }
-    }
+
+    // example: oci://\"diesel\"/diesel@//192.168.2.81:1521/orcl, c.f. sqplus manual
+
+    let splits: Vec<&str> = database_url.split("//").collect();
+    assert_eq!(splits.len(), 3);
+    let userandpw: Vec<&str> = splits[1].split('/').collect();
+    let user = userandpw[0].to_string();
+    let password = unsafe {
+        // discard the @ handle
+        userandpw[1]
+            .slice_unchecked(0, userandpw[1].len() - 1)
+            .to_string()
+    };
+    let db_url = splits[2].to_string();
+
     Ok((user, password, db_url))
 }
 
 impl RawConnection {
+    pub fn check_error(error_handle: *mut ffi::OCIError, status: i32) -> Option<ConnectionError> {
+        let mut errbuf: Vec<u8> = Vec::with_capacity(3072);
```

Comment:
you can't do it like that, oracle won't call `push` in your vector. 

If you want to go this route, you need a Vec with a preallocated size. e.g. `vec![0;3072]` and use `vec.len()` len instead of `vec.capacity()`

Reply:
That actually works because `with_capacity` will allocate a vector with size for n elements. The length then is set using `set_len` in line 145.

Reply:
But why even go down this route (`set_len` is unsafe)? Just use the vector presized which is a lot less confusing.

Reply:
That won't work because we do not know where to set the bound before reading the error message.

Reply:
Sure, you have to resize the errbuf to the byte before first 0 byte. This is what the current version does too, just using a weird way with unsafe functions.

Reply:
I don't see the benefit @geomaniac? If the part with resizing has to be done anyway? How would you do it in safe way? Please notice that `unsafe` is there mainly because of `OCIErrorGet` anyway.

Reply:
`resize` is a safe function in contrast to `set_len`, so nothing bad could possible happen there.

Reply:
The current way with `CStr::from_ptr` and `msg.to_bytes().len()` is just super weird to me.

``` rust
// make sure we always have one 0-byte
let mut buf: Vec<u8> = vec![0; OCI_ERROR_MAXMSG_SIZE2 + 1];
unsafe {
	ffi::OCIErrorGet(
		error_handle as *mut libc::c_void,
		1,
		ptr::null_mut(),
		&mut errcode,
		buf.as_mut_ptr(),
		OCI_ERROR_MAXMSG_SIZE2,
		ffi::OCI_HTYPE_ERROR,
	);
};
let nul_byte_pos = buf.iter().position(|b| *b == 0).expect("we have at least one 0-byte);
buf.resize(nul_byte_pos, 0);
String::from_utf8(buf)
```